Mark deprecated fields in the RoutingRequest

gmellemstrand opened the pull request:
I have now gone through the RoutingRequest and marked all fields that are not used in the Raptor algorithm. Some are not needed anymore while others are not yet implemented.

gmellemstrand commented:
I have now gone through the RoutingRequest and marked all fields that are not used in the Raptor algorithm. Some are not needed anymore while others are not yet implemented.

t2gran commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -67,56 +70,94 @@
+     *
+     * TODO: Needs to  be implemented
+     * */
+    @Deprecated
```
Ideally we should support an enum: DEPART_FROM, ARRIVE_BY, and TIME_TABLE - but I guess we can emulate this by a combination of this flag and setting the from/to times.

t2gran commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -333,7 +420,11 @@
+     * @Deprecated - Not needed in Raptor
+     * */
+    @Deprecated
     public boolean reverseOptimizing = false;
```
Info: The Raptor do have a reverse optimization, but it will not be part of the API to turn it on/off - instead we will have profiles were reverse search will be part of some of them.

gmellemstrand replied:
Even though there is reverse optimization in Raptor, we should probably not use the same parameter, as the functionality is a bit different.

abyrd replied:
Agreed, this should not be controlled from the service API.

t2gran commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -376,13 +476,19 @@
+     * @Deprecated - Not needed in Raptor
      */
+    @Deprecated
     public boolean disableRemainingWeightHeuristic = false;
```
INFO: Heuristic (and reverse search) is part of "profiles" in the Raptor implementation. Profiles are carefully put together to get a working set of parameters that work together. Choosing a profile should be possible from the API, but not setting specific parameters to enable strategies.

t2gran commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -433,7 +539,11 @@
+     * @Deprecated - This will probably be reimplemented
+     * */
+    @Deprecated
     public boolean onlyTransitTrips = false;
```
I am not sure what this is? Is this just a special case of combining different modes?

gmellemstrand replied:
It is used as part of the a star algorithm. To get a variety if itineraries, you ban the trips that were found previously and then to another search. If the first trip you find is walk-only, then you need to ban walk-only in order to not find the same itinerary again. In Raptor this setting is implied and a walk-only search needs to be done separately.

gmellemstrand commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -67,56 +70,94 @@
      * The maximum distance (in meters) the user is willing to walk for access/egress legs.
      * Defaults to unlimited.
+     *
+     * @Deprecated - TODO: This should be supported
```
We do have this as a separate criterion now, only that it is hard-coded. With the current street router we have to limit the search radius when searching for access/egress stops.

gmellemstrand commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -67,56 +70,94 @@
+    /** The set of characteristics that the user wants to optimize for -- defaults to QUICK, or optimize for transit time.
+     *
+     * @Deprecated - Not supported in Raptor
+     * */
```
I think this specific parameter is meant for biking. Available choices are QUICK, SAFE, FLAT, GREENWAYS, TRIANGLE, TRANSFERS (obsolete). If that is the case, it should probably not be mixed in with transit parameters.

abyrd replied:
Many trip planners or trip planning interfaces (OJP for example) still have this assumption that they are performing a single variable optimization. In practice you almost never want the fastest, or flattest, or most extreme path of any kind. You want something like the N near-Pareto-optimal solutions with the lowest generalized cost. 

There are special cases where you want to force the router to do a pure earliest-arrival search. But this field did not originate from such special cases. It's a remnant of the early design of OTP, and reflects a common simplification in user interfaces: the user selects which optimization variable should be weighted the most heavily in the generalized cost.

This is again in a way related to the multiple date format support in the API: it's coming from an early assumption that the trip planning API should map to the user interface components on the front end, with conversion to internal routing parameters happening on the back end. At least those of us working on OTP 2.x seem to favor a new approach: a stricter API reflecting back end routing parameters, with conversion to that normalized representation happening on the front end.

abyrd commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -67,56 +70,94 @@
-    /** Whether maxHours limit should consider wait/idle time between the itinerary and the requested arrive/depart time. */
+    /** Whether maxHours limit should consider wait/idle time between the itinerary and the requested arrive/depart time.
+     *
+     * @Deprecated - Not supported in Raptor
```
I think some parameters you might consider very specific to narrow use cases were added as part of various projects by long-time OTP supporters like TriMet. This is related to the debates on e.g. whether the API should accept many different date formats. It may seem strange from an engineering perspective but it was convenient for certain users.

gmellemstrand commented:
After discussing with @abyrd and @t2gran we came up with another solution for how routing parameters should be handled with AStar/Raptor. This will be in another pull request.

I rebased only the commit that deprecates fields in the RoutingRequest and changed the title of this pull request.

t2gran commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -67,56 +70,95 @@
+    /** Whether the trip should depart at dateTime (false, the default), or arrive at dateTime.
+     *
+     * TODO: Needs to  be implemented
+     * */
```
missing @Deprecated javadoc - is this what we agreed on. The @deprecated tag should be used both in JavaDoc and as a annotation on the method. I see that you are almost consequent, so I am not sure what is best. But, I would just follow the same pattern everywhere  - we will remove the tags when we get back an implement these later.

gmellemstrand replied:
I have made the change now

optionsome commented on src/main/java/org/opentripplanner/routing/core/RoutingRequest.java:
```diff
@@ -246,57 +289,96 @@
     public int otherThanPreferredRoutesPenalty = 300;
 
     /** Set of unpreferred routes for given user. */
+    @Deprecated
```
Why are these unpreferred* parameters deprecated without TODO to implement them? Does implementing them differ from a technical point of view from implementing the preferred* parameters or is it just that they are not in use?

t2gran replied:
This PR needs more JavaDoc, all @Deprecated fields need to be explained why they are deprecated.

t2gran commented:
In my last commit I added more doc to the `RoutingRequest` and `RoutingResource`. Hopefully we can remove some of the parameters from the `RoutingRequest` later, since this is an internal class. I tried to keep both classes up to date and deprecations are added to both.

I am unsure if these parameters are affected:
`transferPenalty, walkReluctance, stairsReluctance, waitReluctance, walkBoardCost, bikeBoardCost`

I assume these are only used in the AStar and not affected:
`turnReluctance, elevatorX, bikeX, carX, waitAtBeginningFactor`

The following parameters are not used in routing - can we delete them? Some is set in the GraphVisualizer and others in the TransmodelIndexGraphQLSchema:
`softWalkLimiting, softPreTransitLimiting, softWalkPenalty, softWalkOverageRate, preTransitPenalty, preTransitOverageRate`

gmellemstrand commented:
@abyrd Could you look over this before I merge?

t2gran commented:
I one of you @optionsome or @abyrd would like to review this, I would appreciate it. I am from the same org. as @gmellemstrand and also helped adding some of the docs, so my review does not count.

optionsome commented:
Isn't use of @Deprecated annotation for things that are not really deprecated but on the TODO list slightly bad practice. However, I'm not sure what would be the ideal solution for this as I think @TODO annotation does not exist out of box.

t2gran commented:
> Isn't use of @Deprecated annotation for things that are not really deprecated but on the TODO list slightly bad practice.
Yes, it is - but is temporary, we will clean this up as part of implementing the weight. I just want the reason why this is not currently working to be as visible as possible - so people do not have to spend time to investigating something that we know is not working.

optionsome commented on src/main/java/org/opentripplanner/api/common/RoutingResource.java:
```diff
@@ -194,36 +244,47 @@
+    /**
+     * The comma-separated list of banned agencies.
+     *
+     * @deprecated TODO OTP2 Regression. Not currently working in OTP2.
```
How do these types of deprecation comments differ from "TODO OTP2: Needs to be implemented"? Is this something that can be reimplemented in the future but no plans to implement it in the near future?

t2gran replied:
I guess you have a point, the plan is to make sure we go trough all `TODO OTP2` before the release of OTP2, but we will clean up the REST API before that. Also, resolving a TODO might not mean fixing it, it can be just to create a new issue here at GitHub
